### PR TITLE
[REEF-62]: Tighten state transition checks in EvaluatorStatusManager

### PR DESCRIPTION
**Pull request by afchung**
This addressed the issue by
  * Checking that the transition of evaluator status is legal.
  * Logging a warning when the transition is illegal.

JIRA: [REEF-62](https://issues.apache.org/jira/browse/REEF-62)

### CODE REVIEW

**Comment by markusweimer** [comment]
The logic looks good. However, we usually have nested `switch` statements for these checks. See `org.apache.reef.runtime.common.evaluator.task.TaskStatus#isLegal` for an example.

In the name of homogeneous code, can you rephrase it that way?

File: lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/driver/evaluator/EvaluatorStatusManager.java [hunk 1]
```diff
@@ -43,8 +43,30 @@ private EvaluatorStatusManager() {
   }
 
   private static boolean isLegal(final EvaluatorState from, final EvaluatorState to) {
-    // TODO
-    return true;
+    switch(from) {
+      case ALLOCATED: {
+        if (to == EvaluatorState.SUBMITTED || to == EvaluatorState.FAILED || to == EvaluatorState.KILLED) {
```
Comment (afchung) [inline lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/driver/evaluator/EvaluatorStatusManager.java]: I'm assuming that an Evaluator with no submitted tasks should never transition to the DONE state. Is this assumption correct?

Comment (markusweimer) [inline lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/driver/evaluator/EvaluatorStatusManager.java]: No. It can transition to `DONE` if it was closed right after allocation or creation of the first context.

**Comment by afchung** [comment]
Are these build failures related to the PR? I was able to build and run tests successfully locally.

**Comment by jwang98052** [comment]
Two test failures seem unrelated to your change. If your local test passed, I guess it is just unstable in those remote connection testing.

**Comment by markusweimer** [comment]
Is this ready to be reviewed again?

**Comment by afchung** [comment]
Yup, please have a look. Thanks!

File: lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/driver/evaluator/EvaluatorStatusManager.java [hunk 2]
```diff
@@ -43,8 +43,34 @@ private EvaluatorStatusManager() {
...
+      return true;
+    }
+
+    if (!isDoneOrFailedOrKilled(from) && isDoneOrFailedOrKilled(to)) {
+      return true;
+    }
+
+    switch(from) {
+      case ALLOCATED: {
+        return to == EvaluatorState.SUBMITTED;
```
Comment (markusweimer) [inline lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/driver/evaluator/EvaluatorStatusManager.java]: It can go from Allocated to any of DONE, FAILED or KILLED.

**Comment by markusweimer** [comment]
I did a pass (see above). Beyond that, I'd still prefer the nested `switch` statements to make things easier to parse.

File: lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/driver/evaluator/EvaluatorStatusManager.java (hunk 2 above)
Comment (afchung) [inline lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/driver/evaluator/EvaluatorStatusManager.java]: Right, the logic is handled here:
!isDoneOrFailedOrKilled(from) && isDoneOrFailedOrKilled(to)

**Comment by afchung** [comment]
In this revision, I used this line:
!isDoneOrFailedOrKilled(from) && isDoneOrFailedOrKilled(to)
to avoid the verbosity of using nested switches, since it seems that any non-terminal state can transition to a terminal state, and a terminal state cannot transition to a different terminal state; other than that, there only seems to be one legal type of state transition for each non-terminal state. Perhaps nested switch statements can provide more flexibility and make the code more clear. I will push another commit. Thanks!

**Comment by afchung** [comment]
This is ready for another look. Thanks!

**Comment by markusweimer** [comment]
Thanks, this looks good, and is much easier to verify :-) I'll test and merge.
